packages/datetime/src/dateRangeInput.tsx

```diff
@@ -430,11 +487,93 @@ export class DateRangeInput extends AbstractComponent<IDateRangeInputProps, IDat
         }
     }
 
+    private getDateRangeForCallback = (currValue?: moment.Moment, currBoundary?: DateRangeBoundary) => {
+        const otherBoundary = this.getOtherBoundary(currBoundary);
+        const otherValue = this.getStateKeysAndValuesForBoundary(otherBoundary).values.selectedValue;
+
+        const currDate = this.getDateForCallback(currValue);
+        const otherDate = this.getDateForCallback(otherValue);
+
+        return (currBoundary === DateRangeBoundary.START)
+            ? [currDate, otherDate]
+            : [otherDate, currDate];
+    }
+
+    private getDateForCallback = (momentDate: moment.Moment) => {
+        if (isMomentNull(momentDate)) {
+            return null;
+        } else if (!momentDate.isValid()) {
+            return new Date(undefined);
+        } else {
+            return fromMomentToDate(momentDate);
+        }
+    }
+
+    private getOtherBoundary = (boundary?: DateRangeBoundary) => {
+        return (boundary === DateRangeBoundary.START) ? DateRangeBoundary.END : DateRangeBoundary.START;
+    }
+
+    private doBoundariesOverlap(boundary: DateRangeBoundary, boundaryDate: moment.Moment) {
+        const otherBoundary = this.getOtherBoundary(boundary);
+        const otherBoundaryDate = this.getStateKeysAndValuesForBoundary(otherBoundary).values.selectedValue;
+
+        // TODO: add handling for allowSingleDayRange
+        if (boundary === DateRangeBoundary.START) {
+            return boundaryDate.isSameOrAfter(otherBoundaryDate);
+        } else {
+            return boundaryDate.isSameOrBefore(otherBoundaryDate);
+        }
+    }
+
     private isControlled = () => {
         return this.props.value !== undefined;
     }
 
+    private isInputEmpty = (inputString: string) => {
+        return inputString == null || inputString.length === 0;
+    }
+
+    private isInputInErrorState = (boundary: DateRangeBoundary) => {
+        const values = this.getStateKeysAndValuesForBoundary(boundary).values;
+        const { isInputFocused, inputString, selectedValue } = values;
+
+        const boundaryValue = (isInputFocused)
+            ? this.dateStringToMoment(inputString)
+            : selectedValue;
+
+        // break down the boolean logic to an elementary level to make it
+        // utterly simple to grok.
+
+        const isEmpty = isMomentNull(boundaryValue);
+        if (isEmpty) {
+            return false;
+        }
+
+        const isValid = boundaryValue.isValid();
+        if (!isValid) {
+            return true;
+        }
+
+        const isInRange = this.isMomentInRange(boundaryValue);
+        if (!isInRange) {
+            return true;
+        }
+
+        // if the boundaries overlap, always consider the END boundary to be
+        // erroneous.
+        const isOverlapping = this.doBoundariesOverlap(boundary, boundaryValue);
+        if (isOverlapping && boundary === DateRangeBoundary.END) {
```

Comment:
I was going back and forth on this. Organized it like this because I didn't want `doBoundariesOverlap` to have any knowledge of the validation semantics - it should just return `true` if the boundaries overlap, as the name implies. By the same token, it felt clearer to keep the `boundary === DateRangeBoundary.END` check inline in both places it appears, because then there's no weird boundary-specific magic that you have to go investigate when reading higher-level code ("but these boundaries _do_ overlap; why is `doBoundariesOverlap` returning false!?")

Maybe different naming could help, but I'm struggling to think of an elegant function name that would include both checks. `isOverlappingEndBoundary` or something?